File: apps/android/app/src/main/java/app/inkuna/android/model/SystemReadingFonts.kt
package app.inkuna.android.model

import android.graphics.fonts.FontStyle
import android.graphics.fonts.SystemFonts
import app.inkuna.core.SystemFontFace
import app.inkuna.core.SystemFontRole

/**
 * Resolves the platform reading faces — Noto Serif for `system-serif`,
 * Roboto for `system-sans` — down to the font files the engine can shape
 * with, from [SystemFonts.getAvailableFonts].
 *
 * Discovery is best-effort by design: a face without a readable file is
 * silently skipped, and the engine substitutes the bundled Noto for that
 * role at selection time. Nothing here throws.
 */
object SystemReadingFonts {

    /**
     * The regular/bold × upright/italic grid for both roles. The platform
     * exposes no family names here, so the family gate is the well-known
     * file names; weight and slant come from each font's declared
     * [FontStyle], never from the name. A variable file enumerated once
     * per named instance collapses to one entry — the core instances
     * weights from its `wght` axis itself.
     */
    fun discover(): List<SystemFontFace> {
        val faces = mutableListOf<SystemFontFace>()
        val seen = mutableSetOf<Triple<String, Int, Boolean>>()
        for (font in SystemFonts.getAvailableFonts()) {
            val file = font.file ?: continue
            val role = roleFor(file.name) ?: continue
            val weight = font.style.weight
            if (weight != FontStyle.FONT_WEIGHT_NORMAL && weight != FontStyle.FONT_WEIGHT_BOLD) continue
            if (!file.canRead()) continue
            val italic = font.style.slant == FontStyle.FONT_SLANT_ITALIC
            if (!seen.add(Triple(file.absolutePath, font.ttcIndex, italic))) continue
            faces += SystemFontFace(
                role = role,
                italic = italic,
                weight = weight.toUShort(),
                filePath = file.absolutePath,
                // The engine opens the face by its collection index; the
                // name scan is iOS's path, where no index is exposed.
                postScriptName = null,
                ttcHint = font.ttcIndex.toUInt(),
            )
        }
        return faces
    }

    /**
     * Latin core families only. The per-script Notos (NotoSerifThai-…,
     * NotoSansCJK-…) belong to the platform's fallback chain, not to the
     * reading role, so the patterns anchor the whole base name.
     */
    private fun roleFor(fileName: String): SystemFontRole? = when {
        SERIF_FILE.matches(fileName) -> SystemFontRole.SERIF
        SANS_FILE.matches(fileName) -> SystemFontRole.SANS
        else -> null
    }

    private val SERIF_FILE = Regex("NotoSerif-\\w+\\.(ttf|otf)", RegexOption.IGNORE_CASE)
    private val SANS_FILE = Regex("Roboto-\\w+\\.(ttf|otf)", RegexOption.IGNORE_CASE)
}

File: apps/ios/Inkuna/Reader/Engine/SystemReadingFonts.swift
import CoreText
import UIKit

/// Resolves the platform reading faces — New York for `system-serif`,
/// San Francisco for `system-sans` — down to the font files the engine
/// can shape with.
///
/// Discovery is best-effort by design: a face whose file cannot be
/// resolved or read is silently skipped, and the engine substitutes the
/// bundled Noto for that role at selection time. Nothing here throws.
enum SystemReadingFonts {
    /// The regular/bold × upright/italic grid for both roles, deduplicated.
    /// Both weights of a variable face resolve to the same file; the core
    /// instances weights from the `wght` axis itself, so one entry per
    /// distinct face is enough.
    static func discover() -> [SystemFontFace] {
        var faces: [SystemFontFace] = []
        var seen: Set<FaceKey> = []
        // 400 before 700, so a variable file keeps its regular entry.
        for (weight, uiWeight) in [(UInt16(400), UIFont.Weight.regular), (UInt16(700), .bold)] {
            for italic in [false, true] {
                for (role, design) in [
                    (SystemFontRole.serif, UIFontDescriptor.SystemDesign.serif),
                    (SystemFontRole.sans, .default),
                ] {
                    guard let face = resolve(
                        role: role,
                        design: design,
                        weight: weight,
                        uiWeight: uiWeight,
                        italic: italic
                    ) else { continue }
                    let key = FaceKey(
                        role: role,
                        italic: italic,
                        filePath: face.filePath,
                        postScriptName: face.postScriptName
                    )
                    if seen.insert(key).inserted {
                        faces.append(face)
                    }
                }
            }
        }
        return faces
    }

    private struct FaceKey: Hashable {
        let role: SystemFontRole
        let italic: Bool
        let filePath: String
        let postScriptName: String?
    }

    private static func resolve(
        role: SystemFontRole,
        design: UIFontDescriptor.SystemDesign,
        weight: UInt16,
        uiWeight: UIFont.Weight,
        italic: Bool
    ) -> SystemFontFace? {
        var descriptor = UIFont.systemFont(ofSize: 17, weight: uiWeight).fontDescriptor
        if design != .default {
            // A missing design means the platform has no such face.
            guard let designed = descriptor.withDesign(design) else { return nil }
            descriptor = designed
        }
        if italic {
            guard let slanted = descriptor.withSymbolicTraits(.traitItalic) else { return nil }
            descriptor = slanted
        }

        let font = CTFontCreateWithFontDescriptor(descriptor as CTFontDescriptor, 17, nil)
        // An italic request the platform satisfied with the upright face
        // must not be registered as italic, or it would hijack emphasis.
        if italic, !CTFontGetSymbolicTraits(font).contains(.traitItalic) {
            return nil
        }
        guard
            let url = CTFontCopyAttribute(font, kCTFontURLAttribute) as? URL,
            url.isFileURL
        else { return nil }
        let path = url.path
        // System font files live under /System and are sandbox-readable;
        // anything the process cannot actually read is skipped up front so
        // the engine never records a face it cannot open.
        guard FileManager.default.isReadableFile(atPath: path) else { return nil }

        // Only a collection needs the PostScript name: the core scans the
        // .ttc for the matching face. For single-face files the name CTFont
        // reports can be an alias of the name table's, and a mismatch would
        // reject a perfectly good face — index 0 is already exact.
        let postScriptName: String? = url.pathExtension.lowercased() == "ttc"
            ? CTFontCopyPostScriptName(font) as String
            : nil

        return SystemFontFace(
            role: role,
            italic: italic,
            weight: weight,
            filePath: path,
            postScriptName: postScriptName,
            ttcHint: nil
        )
    }
}
